src/ChunkedStreamDecoder.php

```diff
@@ -0,0 +1,147 @@
+<?php
+
+namespace React\HttpClient;
+
+use Evenement\EventEmitterTrait;
+use Exception;
+use React\Stream\ReadableStreamInterface;
+use React\Stream\Util;
+use React\Stream\WritableStreamInterface;
+
+class ChunkedStreamDecoder implements ReadableStreamInterface
+{
+    const CRLF = "\r\n";
+
+    use EventEmitterTrait;
+
+    /**
+     * @var string
+     */
+    protected $buffer = '';
+
+    /**
+     * @var int
+     */
+    protected $remainingLength = 0;
+
+    /**
+     * @var bool
+     */
+    protected $nextChunkIsLength = true;
+
+    /**
+     * @var ReadableStreamInterface
+     */
+    protected $stream;
+
+    /**
+     * @param ReadableStreamInterface $stream
+     */
+    public function __construct(ReadableStreamInterface $stream)
+    {
+        $this->stream = $stream;
+        $this->stream->on('data', array($this, 'handleData'));
+        Util::forwardEvents($this->stream, $this, [
+            'error',
+        ]);
+    }
+
+    /** @internal */
+    public function handleData($data)
+    {
+        $this->buffer .= $data;
+
+        do {
+            $bufferLength = strlen($this->buffer);
+            $continue = $this->iterateBuffer();
+            $iteratedBufferLength = strlen($this->buffer);
+        } while (
+            $continue &&
+            $bufferLength !== $iteratedBufferLength &&
+            $iteratedBufferLength > 0 &&
+            strpos($this->buffer, static::CRLF) !== false
+        );
+    }
+
+    protected function iterateBuffer()
+    {
+        if ($this->nextChunkIsLength) {
+            $crlfPosition = strpos($this->buffer, static::CRLF);
+            if ($crlfPosition === false) {
+                return false; // Chunk header hasn't completely come in yet
+            }
+            $this->nextChunkIsLength = false;
+            $lengthChunk = substr($this->buffer, 0, $crlfPosition);
+            if (strpos($lengthChunk, ';') !== false) {
+                list($lengthChunk) = explode(';', $lengthChunk, 2);
+            }
+            $lengthChunk = trim($lengthChunk, "\r\n");
+            if (!ctype_xdigit($lengthChunk)) {
```

Comment:
> What is the point in trimming this here?

That shouldn't have gotten in, fixing that now. The trimming was because some tests use str_split to send the test string in character for character and that caused some `\r` and `\n` to stay in front of the header for some reason. Thought I had fixed that :disappointed: 

> How about checking dechex(hexdec($original)) === $original and within reasonable limits?

Yup that works great, set the limit to `32KB`, which while is a big chunk is still within reasonable range. But if you consider that to high we can always change it. Also I'm constructing the class in such a way a user can specify it's own limit.